### PR TITLE
Fix authenticated media

### LINKED ISSUE #685: NetworkAccessManager doesn't work with authenticating media servers
NAM doesn't know anything about access tokens and converts `mxc` URLs into unauthenticated Matrix requests - failing on media servers that enforce authenticated access.

### PR DESCRIPTION
**Pull request by TobiasFella**

### CODE REVIEW

**Review by KitsuneRal — CHANGES_REQUESTED** [review]
I understand the problem but that's not a good way to fix it, unfortunately: `NAM::createRequest()` can be called from several threads (in QML context, in particular), and `Connection` is not a thread-safe object. There's also a conceptual problem in that the library so far managed to stay somewhat layered (NAM - ConnectionData+BaseJob - Connection/Room), and the `Connection` dependency pierces through the layers.

Why not store access tokens in `d` instead?

**Comment by TobiasFella** [comment]
Yeah, pulling the Connection in there is a bit terrible. I did it this way because then we're already prepared for a future where access tokens change at runtime (refresh tokens, OIDC), but i guess we do need to push the token in there directly

**Comment by KitsuneRal** [comment]
The control point for tokens (`ConnectionData::setToken()`) is already there anyway, and `ConnectionData` can push the token down to `NAM`, every time the token is updated. It can even store the token in `NAM`, with `ConnectionData::accessToken()` retrieving it from there (with tiny loss of efficiency vs. tiny win of memory).

**Comment by TobiasFella** [comment]
right, that's nicer. I've updated the patch

File: Quotient/networkaccessmanager.cpp [hunk 1]
```diff
@@ -25,6 +25,7 @@ class {
     struct ConnectionData {
         QString accountId;
         HomeserverData hsData;
+        QByteArray accessToken;
```
Comment (KitsuneRal) [inline Quotient/networkaccessmanager.cpp]: I'm looking at `HomeserverData` usages now and get to wonder whether we should add `accessToken` to it instead. This structure is meant to encapsulate all data necessary to form a network request to Matrix API (I should have documented it...), and the access token fits that bill perfectly. The `NAM` code could then be simplified a bit further.

From a paranoid standpoint, whoever calls `Connection::homeserverData()` would then get everything necessary to make an _authenticated_ request without thinking of it. However, that method is in a protected section, next to `connectionData()`, and calling `connectionData()->accessToken()` would get exactly the same thing and is available today to exactly the same callers. And one still has to explicitly set the header with the access token.

Comment (TobiasFella) [reply Quotient/networkaccessmanager.cpp]: done

File: Quotient/networkaccessmanager.cpp [hunk 2]
```diff
@@ -171,6 +179,7 @@ QNetworkReply* NetworkAccessManager::createRequest(
     // Convert mxc:// URL into normal http(s) for the given homeserver
     QNetworkRequest rewrittenRequest(request);
     rewrittenRequest.setUrl(DownloadFileJob::makeRequestUrl(hsData, url));
+    rewrittenRequest.setRawHeader("Authorization", QByteArrayLiteral("Bearer ") + d.getConnection(accountId).accessToken);
```
Comment (KitsuneRal) [inline Quotient/networkaccessmanager.cpp]:
```suggestion
    rewrittenRequest.setRawHeader("Authorization", QByteArrayLiteral("Bearer ") + hsData.accessToken);
```

**Review by KitsuneRal — APPROVED** [review]

**Pull request merged**